avocado/plugins/spawners/remote.py

```diff
@@ -0,0 +1,210 @@
+import asyncio
+import contextlib
+import json
+import logging
+import os
+import shlex
+
+from aexpect import remote
+
+from avocado.core.plugin_interfaces import Init, Spawner
+from avocado.core.settings import settings
+from avocado.core.spawners.common import SpawnerMixin, SpawnMethod
+
+LOG = logging.getLogger(__name__)
+
+
+class RemoteSpawnerException(Exception):
+    """Errors more closely related to the spawner functionality"""
+
+
+class RemoteSpawnerInit(Init):
+
+    description = "Remote (host) based spawner initialization"
+
+    def initialize(self):
+        section = "spawner.remote"
+
+        help_msg = "List of already available remote host slots to spawn in"
+        settings.register_option(
+            section=section, key="slots", help_msg=help_msg, key_type=list, default=[]
+        )
+
+        help_msg = "Remote host setup hook command to customize optional new hosts"
+        settings.register_option(
+            section=section, key="setup_hook", help_msg=help_msg, default=""
+        )
+
+
+def with_slot_reservation(fn):
+    """
+    Decorator for slot cache context manager.
+
+    :param fn: function to run with slot reservation
+    :type fn: function
+    :returns: same function with the slot now reserved
+    :rtype: function
+
+    The main reason for the decorator is to not have to indent the entire
+    task running function in order to safely release the slot upon any error.
+    """
+
+    async def wrapper(self, runtime_task):
+        with RemoteSpawner.reserve_slot(self, runtime_task) as slot:
```

Comment:
This is identical (besides the "RemoteSpawner" reference that can become a variable) to the version on the lxc spawner.  Maybe come up with a version that suits both and share it?

Reply:
Yes, I guess we could refactor a shared mechanism for using "available slots" for these types of spawners. Will take a look.